src/views/ApartmentApp/components/StudentApplication/index.js

```diff
@@ -0,0 +1,794 @@
+import SimpleSnackbar from '../../../../components/Snackbar';
+import ApartmentHeader from './components/ApartmentHeader';
+import InstructionsCard from './components/InstructionsCard';
+import ApplicationDataTable from './components/ApplicationDataTable';
+import ApplicantList from './components/ApplicantList';
+import HallSelection from './components/HallSelection';
+import Agreements from './components/Agreements';
+import SaveButton from './components/SaveButton';
+import housing from '../../../../services/housing';
+import user from '../../../../services/user';
+const MAX_NUM_APPLICANTS = 8;
+
+/**
+ * @typedef { import('../../../../services/user').StudentProfileInfo } StudentProfileInfo
+ * @typedef { import('../../../../services/housing').ApartmentApplicant } ApartmentApplicant
+ * @typedef { import('../../../../services/housing').ApartmentChoice } ApartmentChoice
+ * @typedef { import('../../../../services/housing').ApplicationDetails } ApplicationDetails
+ */
+
+/**
+ * Renders the page for the student apartment application
+ * @param {Object} props The React component props
+ * @param {*} props.authentication The user authentication
+ * @param {StudentProfileInfo} props.userProfile The student profile info of the current user
+ * @returns {JSX.Element} JSX Element for the student application web page
+ */
+const StudentApplication = ({ userProfile, authentication }) => {
+  const [loading, setLoading] = useState(true);
+  const [saving, setSaving] = useState(false);
+  const [unsavedChanges, setUnsavedChanges] = useState(false);
+
+  /** @type {[ApplicationDetails, React.Dispatch<React.SetStateAction<ApplicationDetails>>]} */
+  const [applicationDetails, setApplicationDetails] = useState({
+    ApplicationID: null,
+    DateSubmitted: null,
+    DateModified: null,
+    EditorUsername: null,
+    EditorEmail: null,
+    Applicants: [],
+    ApartmentChoices: [],
+  });
+  /** @type {[Number, React.Dispatch<React.SetStateAction<Number>>]} */
+  const [applicationID, setApplicationID] = useState(null); // Default value of -1 indicate to backend that the application ID number is not yet known
+  const [dateSubmitted, setDateSubmitted] = useState(null); // The date the application was submitted, or null if not yet submitted
+  const [dateModified, setDateModified] = useState(null); // The date the application was last modified, or null if not yet saved/modified
+  const [editorUsername, setEditorUsername] = useState(null); // The username of the application editor
+  /** @type {[ApartmentApplicant[], React.Dispatch<React.SetStateAction<ApartmentApplicant[]>>]} Array of applicant info */
+  const [applicants, setApplicants] = useState([]);
+  /** @type {[ApartmentChoice[], React.Dispatch<React.SetStateAction<ApartmentChoice[]>>]} Array of apartment choice info */
+  const [preferredHalls, setPreferredHalls] = useState([]); // Properties 'HallName' and 'HallRank' must be capitalized to match the backend
```

Comment:
I'm very curious why you have separate state variables here for things that are part of the `applicationDetails` state. Keeping two copies of the same data is the easiest way of getting broken state that behaves in unpredictable ways. Why not just use the `applicationDetails` state itself?

Reply:
I actually plan to completely replace the separate state variables with that `applicationDetails` state like you described because I realized it is beneficial to the reliability of other new features, and IMO it makes the code logic more intuitive to read. I just haven't gotten around to fully implementing that change yet

Reply:
When you work on that, it might be worth considering `useReducer`. That hook is much more simple to use with complex state objects than `useState`

Reply:
Interesting, I will read up on that right now

Work on this is in progress at #1072